internal/hcsoci/drivers.go

```diff
@@ -0,0 +1,115 @@
+	"github.com/pkg/errors"
+)
+
+// getAssignedDeviceUtilityTools gets the path to device-util on the server host
+func getAssignedDeviceUtilityTools(coi *createOptionsInternal) (string, error) {
+	tools, ok := coi.Spec.Annotations[oci.AnnotationAssignedDeviceUtilityTools]
+	if !ok || tools == "" {
+		return "", fmt.Errorf("no driver tools were specified %s", tools)
+	}
+	if _, err := os.Stat(tools); err != nil {
+		return "", errors.Wrapf(err, "failed to find device installation tools at %s", tools)
+	}
+	return tools, nil
+}
+
+// getAssignedDeviceKernelDrivers gets any device drivers specified on the spec
+func getAssignedDeviceKernelDrivers(coi *createOptionsInternal) ([]string, error) {
+	csDrivers, ok := coi.Spec.Annotations[oci.AnnotationAssignedDeviceKernelDrivers]
+	if !ok || csDrivers == "" {
+		return nil, nil
+	}
+	drivers := strings.Split(csDrivers, ",")
+	for _, driver := range drivers {
+		if _, err := os.Stat(driver); err != nil {
+			return nil, errors.Wrapf(err, "failed to find path to drivers at %s", driver)
+		}
+	}
+	return drivers, nil
+}
+
+// setupDeviceUtilTools finds the utility tools' host path, mounts it using vsmb, and
+// returns the UVM path to the tools
+func setupDeviceUtilTools(ctx context.Context, coi *createOptionsInternal, r *Resources) (string, error) {
+	toolsHostPath, err := getAssignedDeviceUtilityTools(coi)
+	if err != nil {
+		return "", err
+	}
+	resultUVMPaths, err := mountDirectoriesUVM(ctx, coi.HostingSystem, r, toolsHostPath)
+	if err != nil {
+		return "", nil
+	}
+	return resultUVMPaths[0], nil
+}
+
+// setupDriversWindows finds specified kernel driver directories, mounts them using vsmb,
+// then installs them in the UVM
+func setupDriversWindows(ctx context.Context, coi *createOptionsInternal, resources *Resources) error {
+	drivers, err := getAssignedDeviceKernelDrivers(coi)
+	if err != nil {
+		return err
+	}
+	if drivers == nil {
+		// no drivers were specified, skip installing drivers
+		return nil
+	}
+	driverUVMPaths, err := mountDirectoriesUVM(ctx, coi.HostingSystem, resources, drivers...)
+	if err != nil {
+		return err
+	}
+	return execPnPInstallAllDrivers(ctx, coi.HostingSystem, driverUVMPaths)
+}
+
+// mountDirectoriesUVM mounts all specified driver files using VSMB and returns their path
+// in the UVM
+func mountDirectoriesUVM(ctx context.Context, vm *uvm.UtilityVM, r *Resources, hostPaths ...string) (resultUVMPaths []string, err error) {
+	for _, d := range hostPaths {
+		options := &hcsschema.VirtualSmbShareOptions{
+			ReadOnly:            true,
+			CacheIo:             true,
+			ShareRead:           true,
+			ForceLevelIIOplocks: true,
+			NoDirectmap:         vm.DevicesPhysicallyBacked(),
+		}
+		share, err := vm.AddVSMB(ctx, d, "", options)
+		if err != nil {
+			return nil, fmt.Errorf("failed to add VSMB share to utility VM for path %+v: %s", d, err)
+		}
+		r.resources = append(r.resources, share)
+		uvmPath, err := vm.GetVSMBUvmPath(ctx, d)
```

Comment:
(This doesnt have to be added here, I'll likely add it to the mapped dir shimdiag work so more of a heads up) This feels wrong how we have to call a uvm method when the information we need is already located in the vsmb share struct, we should just export a method on VSMBShare called UvmPath so that we can just call 'share.UvmPath()'.

Reply:
https://github.com/microsoft/hcsshim/blob/70248d045a858ae21b83a043fa80ee6b1a47a84a/internal/uvm/vsmb.go#L37 can use this now 😄